Models (Data Mapper, Repository, Entity)

aespasa opened the pull request.

aespasa committed:
Work on ORM and DBAL

This commit is a work in progress of Repository pattern, Entity Manager, Entities, Data Mapping and Unit of Work

TBoileau commented on src/Database/Adapter/MySQLAdapterInterface.php:
```diff
@@ -0,0 +1,21 @@
+
+interface MySQLAdapterInterface extends AdapterInterface
+{
+    /** Connect to a PDO instance
+     * @return PDO
+     */
+    public function connect(): PDO;
+
+    /** Execute a SQL prepared query
+     * @param string $rawQuery
+     * @param array<string, string> $bind
+     * @return PDOStatement|false
+     */
```
Pourquoi ces méthodes sont ici et pas dans AdapterInterface, pas sur que cette interface soit utile d'ailleurs.
On ne devrait avoir qu'une seule interface et X implémntations si besoin

aespasa replied:
J'ai remonté ces méthodes dans l'AdapterInterface et supprimé le MySQLInterface, j'ai mis un return mixed sur le connect, j'ai juste un doute pour la méthode query() car elle prends en paramètre un string $rawquery et une array $bind pour faire des requêtes préparés mais ca a l'air très spécifique au SQL ce genre de chose non ? Si je souhaite un jour mettre autre chose est ce que le bind sera toujours pertinent ?

TBoileau replied:
Non mais dans TOUS les cas tu gères avec PDO. Ce qui va changer c'est juste l'url de connexion que tu gères déjà dans MysqlAdapter

TBoileau commented on src/Database/DataMapper.php:
```diff
@@ -0,0 +1,13 @@
+
+class DataMapper implements MapperInterface
+{
+    public function mapRowToEntity(array $data, EntityInterface $entity)
+    {
+
+    }
```
Hum ça sert à quoi ? Car un mapper, ne sert qu'à récupérer le mapping d'une entité en mode flat, en gros juste un moyen de récupérer ses infos
La j'ai l'impression que tu veux hydrater quelques chose

TBoileau commented on src/Entity/EntityInterface.php:
```diff
@@ -0,0 +1,8 @@
+
+namespace Blog\Entity;
+
+interface EntityInterface
```
A quoi te sert cette interface ?

TBoileau commented on src/Entity/Post.php:
```diff
@@ -0,0 +1,42 @@
+
+namespace Blog\Entity;
+
+class Post extends EntityManager implements EntityInterface
```
Heu pourquoi une entité étant un EntityManager, l'entity manager est un service

TBoileau replied:
Une entité ne devrait étendre rien, si ce n'est une autre entité, et il ne devrait pas non plus avoir d'interface

aespasa replied:
Oui pour l'EM c'est moche, mais pour l'EntityInterface je me dis que c'est surtout pour pouvoir typer mes méthodes de mon Entity Manager, par exemple $em->persist(EntityInterface $entity), sinon comment faire pour s'assurer que l'on attends bien une entité sans classe parente ou interface, selon l'entité que l'on lui passe ? (User, Post..)

TBoileau replied:
Parce que tu vas vérifier avec le data mapper si c'est bien une entité

TBoileau commented on src/Repository/Repository.php:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * @return string
+     */
+    public function generateId(): string
+    {
+        return Uuid::uuid4()->toString();
+    }
```
Ce n'est pas responsabilité du repository de générer un Uuid

aespasa replied:
C'est quoi c'est un service à part ou peut être l'entity manager ?

TBoileau replied:
ça peut être un service à part ou simplement utilisé de manière native

TBoileau commented on src/Repository/Repository.php:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * @return AdapterInterface
+     */
+    public function getAdapter(): AdapterInterface
+    {
+        return $this->adapter;
+    }
+
+    /**
+     * @param AdapterInterface $adapter
+     * @return Repository
+     */
+    public function setAdapter(AdapterInterface $adapter): Repository
+    {
+        $this->adapter = $adapter;
+        return $this;
+    }
```
Il ne devrait pas avoir d'adapter. On injecte seulement l'entité manager à l'intérieur

aespasa replied:
Ce n'est pas l'entitymanager qui est censé appeler le repository ?

TBoileau replied:
Non du tout, le repository est un service à laquelle tu injectes l'entity manager

TBoileau commented on src/Database/DataMapper.php:
```diff
@@ -0,0 +1,56 @@
+    /** @var array<mixed>  */
+    private array $mapping = [];
+
+    public function mapEntity(object $entity): array
```
Remplacer par un string (qui sera composer du FQCN)

TBoileau commented on src/Database/DataMapper.php:
```diff
@@ -0,0 +1,56 @@
+
+    public function mapEntity(object $entity): array
+    {
+        $reflClass = new \ReflectionClass($entity);
+
+        $classAttributesName = [];
+        foreach ($reflClass->getAttributes() as $attribute) {
+            $classAttributesName[] = $attribute->getName();
+        }
+
+        if (!in_array(Entity::class, $classAttributesName)) {
+            throw new \Exception('Attribute ' . Entity::class . ' is missing from ' . $reflClass->getName());
+        }
+
+        if (!in_array(Table::class, $classAttributesName)) {
+            throw new \Exception('Attribute ' . Table::class . ' is missing from ' . $reflClass->getName());
+        }
+
+        foreach ($reflClass->getAttributes() as $attribute) {
+            $args = $attribute->getArguments();
+
+            foreach ($args as $key => $arg) {
+                $this->$key = $arg;
+            }
+        }
+
+        foreach ($reflClass->getProperties() as $property) {
+            foreach ($property->getAttributes() as $attribute) {
+                $this->mapping[$attribute->getName()][] = $attribute->getArguments();
+            }
+        }
+
+        return [
+            'tableName' => $this->tableName,
+            'repositoryClass' => $this->repositoryClass,
+            'data ' => $this->mapping
+        ];
```
Stocker le mapping quelque part, et ainsi éviter de refaire à chaque fois le même traitement, alors que tu aurais supposément déjà fait un mapping de cette même entité

TBoileau commented on src/Entity/EntityManager.php:
```diff
@@ -0,0 +1,68 @@
+        $this->unitOfWork = new UnitOfWork($this);
+    }
+
+    /**
+     * @return AdapterInterface
+     */
+    public function getAdapter(): AdapterInterface
+    {
+        return $this->adapter;
+    }
+
+    /**
+     * @param AdapterInterface $adapter
+     * @return EntityManager
+     */
+    public function setAdapter(AdapterInterface $adapter): self
+    {
+        $this->adapter = $adapter;
+        return $this;
+    }
```
Pourquoi avoir besoin d’assesseurs pour l'adapter alors que tu as un getConnection` plus bas? Je pense que c'est inutile.

TBoileau commented on src/Entity/Post.php:
```diff
@@ -0,0 +1,122 @@
+
+class Post
+{
+    /**
+     * @var string
+     */
```
La PHPDoc est inutile si tu utilises le typage

TBoileau commented on src/ORM/UnitOfWork.php:
```diff
@@ -0,0 +1,58 @@
+
+    }
+
+    public function persist(object $entity)
+    {
+        $oid = spl_object_id($entity);
+
+
+    }
```
Inutile

TBoileau commented on src/ORM/UnitOfWork.php:
```diff
@@ -0,0 +1,58 @@
+
+    }
+
+    public function commit(object $entity): void
+    {
+
+    }
```
Cette méthode va exécuter les requêtes (insertions  modifications, suppression).
Pas besoin d'argument

aespasa committed:
Work in progress on Models

aespasa committed:
Delete node_modules

TBoileau commented on src/Attribute/Column.php:
```diff
@@ -0,0 +1,19 @@
+{
+    public function __construct(
+        private string $name,
+        private string $type = 'string',
```
Utiliser enum

TBoileau commented on src/Attribute/Column.php:
```diff
@@ -0,0 +1,19 @@
+final class Column
+{
+    public function __construct(
+        private string $name,
+        private string $type = 'string',
+        private bool $nullable = false,
+        private bool $unique = false
```
Mettre en publique

aespasa committed:
update gitignore

TBoileau commented on src/Attribute/Entity.php:
```diff
@@ -0,0 +1,16 @@
+final class Entity
+{
+    public function __construct(
+        protected string $repositoryClass
```
Pourquoi protected ? => publique

TBoileau commented on src/Attribute/Table.php:
```diff
@@ -0,0 +1,16 @@
+final class Table
+{
+    public function __construct(
+        private string $tableName
```
Mettre en public

TBoileau commented on src/Database/DataMapper.php:
```diff
@@ -0,0 +1,78 @@
+        $reflClass = new \ReflectionClass($entity);
+
+        $classAttributesName = [];
+        foreach ($reflClass->getAttributes() as $attribute) {
+            $classAttributesName[] = $attribute->getName();
+        }
+
+        if (!in_array(Entity::class, $classAttributesName)) {
+            throw new \Exception('Attribute ' . Entity::class . ' is missing from ' . $reflClass->getName());
+        }
+
+        if (!in_array(Table::class, $classAttributesName)) {
+            throw new \Exception('Attribute ' . Table::class . ' is missing from ' . $reflClass->getName());
+        }
+
+        foreach ($reflClass->getAttributes() as $attribute) {
+            $args = $attribute->getArguments();
+
+            foreach ($args as $key => $arg) {
+                $this->{$key} = $arg;
+            }
+        }
+
+        $mapping = [];
+        foreach ($reflClass->getProperties() as $property) {
+            foreach ($property->getAttributes() as $attribute) {
+                $mapping[] = array_merge(
+                    $attribute->getArguments(),
+                    ['propertyName' => $property->getName()]
+                );
+            }
+        }
+
+        $this->entities->offsetSet($entity, [
+            'tableName' => $this->tableName,
+            'repositoryClass' => $this->repositoryClass,
+            'columns' => $mapping
+        ]);
+
+        return $this->entities->offsetGet($entity);
```
Tu te compliques la viee, tu peux directement recupérer un instance de tes classe de mapping : https://www.php.net/manual/fr/reflectionattribute.newinstance.php

TBoileau replied:
Au lieu de renvoyer un tableau avec les instance de ton mapping, implémenter une classe Metadata, qui comprend :

- class => FQCN
- entity
- table
- id
- columns

TBoileau commented on src/Database/MapperInterface.php:
```diff
@@ -0,0 +1,12 @@
+     * @param string $entity
+     * @return array<string>
+     */
+    public function mapEntity(string $entity): array;
```
Je l'appelerais "resolve"

TBoileau commented on src/ORM/Persister/EntityPersister.php:
```diff
@@ -0,0 +1,22 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Blog\ORM\Persister;
+
+use Blog\Database\Adapter\AdapterInterface;
+use Blog\Entity\EntityManager;
+
+class EntityPersister
+{
+    public function __construct(
+        private EntityManager $entityManager,
+        private AdapterInterface $adapter
+    ) {
+    }
+
+    public function executeInserts(array $entities, array $mapping)
+    {
+        dump($entities);
+    }
+}
```
Useless

aespasa committed:
Work in progress on models

executeInsertions method works, and we can insert some entities with $em->add($entity) and $em->flush()

The DataMapper read the attached attributes of an entity and create a Metadata instance through its getMetadata() method.

The UnitOfWork executes the queued insertions with an SQL transaction and rollback changes if an error occurs.

TBoileau commented on src/Database/Adapter/MySQLAdapter.php:
```diff
@@ -72,25 +60,41 @@ public function query(string $rawQuery, array $bind = []): PDOStatement|false
-                $statement->execute($query[1] ?? []);
+            $rowCount = 0;
+            foreach ($this->transactions as $transaction) {
+                $query = $transaction[0];
+                $bind = $transaction[1];
```
Pas très fan de ça, soit créer une classe simple ou utiliser un tableau associatif :
```php
/**
 * @var array<array-key, array{query: string, bind: array}>
 */
private array $transactions = [];

```

TBoileau commented on src/Database/DataMapper.php:
```diff
@@ -3,76 +3,102 @@
-                $this->{$key} = $arg;
+        $metadatas->setFqcn($fqcn);
+
+        foreach ($reflAttribute as $attribute) {
+            $attributeInstance = $attribute->newInstance();
+
+            switch ($attribute->getName()) {
+                case Entity::class:
+                    /** @var Entity $attributeInstance */
+                    $metadatas->setEntity($attributeInstance);
+                    break;
+                case Table::class:
+                    /** @var Table $attributeInstance */
+                    $metadatas->setTable($attributeInstance);
+                    break;
+                default:
+                    throw new Exception('Unhandled attribute ' . $attribute->getName());
```
Je ne ferais pas de boucle, utiliser le première paramètre de getAttributes

aespasa replied:
J'aimerai le laisser car ça me permet de voir s'il y a un attribut non prévu facilement et il faut  quand même créer les instances, ça reste c'est assez verbeux : 

```php
$tableAttributeInstance = ($reflClass->getAttributes(Table::class))->newInstance();
$metadatas->setTable($reflClass->getAttributes(Table::class));
```

Qu'est ce que tu en pense ?

TBoileau commented on src/ORM/UnitOfWork.php:
```diff
@@ -4,68 +4,63 @@
+            foreach ($mapping->getColumns() as $column) {
+                $bind[':' . $column->name] = match ($column->type) {
+                    // @phpstan-ignore-next-line
+                    Type::HTML => htmlspecialchars($object->{'get' . ucfirst($column->propertyName)}()),
```
ne jamais faire ça en entrée, l'échappement du html se fait toujours à l'affichage

aespasa committed:
Implement findAll() method, implement Hydration and refactor some parts of code

aespasa committed:
Creating an Hydrator, automatic entity's ids generation with the AbstractEntity class, moving all ORM related files to the ORM folder.

We can now get an entity id through $entity->getId() before it was saved into the database, in that way we can associate multiple models without doing an $em->flush() between each different entity persists.

aespasa committed:
Add find() and findOneBy() methods

aespasa committed:
Add deletion, refactor some code and fix some bugs

TBoileau commented on src/Entity/AbstractEntity.php:
```diff
@@ -0,0 +1,30 @@
+<?php
+
+namespace Blog\Entity;
+
+use Blog\ORM\Mapping\Attribute\Id;
+use Ramsey\Uuid\Uuid;
+
+abstract class AbstractEntity
+{
+    #[Id()]
+    protected string $id;
+
+    public function __construct()
+    {
+        if (!isset($this->id)) {
+            $this->id = (string)Uuid::uuid4();
+        }
+    }
+
+    public function getId(): string|false
+    {
+        return $this->id ?? false;
+    }
+
+    public function setId(string $id): self
+    {
+        $this->id = $id;
+        return $this;
+    }
+}
```
Pas fan de cette classe abstraite

TBoileau commented on src/ORM/Hydration/AbstractHydrator.php:
```diff
@@ -0,0 +1,15 @@
+        protected DataMapper $mapper
+    ) {
+    }
+}
```
Inutile

aespasa committed:
Add entity updates to the UnitOfWork

acecconato commented:
close #8 
close #9
